source/beamline/SXRMB/SXRMBBeamline.cpp

```diff
@@ -128,148 +93,210 @@ AMMotorGroupObject *SXRMBBeamline::microprobeSampleStageMotorGroupObject() const
 	return motorGroup_->motorGroupObject("Microprobe Stage - X, Z, Y");
 }
 
-AMReadOnlyPVControl* SXRMBBeamline::beamlineStatus() const {
+AMReadOnlyPVControl* SXRMBBeamline::beamlineStatus() const
+{
 	return beamlineStatus_;
 }
 
-bool SXRMBBeamline::isConnected() const{
+bool SXRMBBeamline::isConnected() const
+{
 	return energy_->isConnected() && beamlineStatus_->isConnected() && microprobeSampleStageControlSet_->isConnected();
 }
 
-CLSBasicScalerChannelDetector* SXRMBBeamline::i0Detector() const{
+CLSBasicScalerChannelDetector* SXRMBBeamline::i0Detector() const
+{
 	return i0Detector_;
 }
 
-CLSBasicScalerChannelDetector* SXRMBBeamline::teyDetector() const{
+CLSBasicScalerChannelDetector* SXRMBBeamline::teyDetector() const
+{
 	return teyDetector_;
 }
 
-AMBasicControlDetectorEmulator* SXRMBBeamline::energyFeedbackDetector() const{
+AMBasicControlDetectorEmulator* SXRMBBeamline::energyFeedbackDetector() const
+{
 	return energyFeedbackDetector_;
 }
 
-SXRMBBrukerDetector* SXRMBBeamline::brukerDetector() const{
+SXRMBBrukerDetector* SXRMBBeamline::brukerDetector() const
+{
 	return brukerDetector_;
 }
 
-AMAction3* SXRMBBeamline::createBeamOnActions() const{
+AMAction3* SXRMBBeamline::createBeamOnActions() const
+{
 	if(!isConnected())
 		return 0;
 
-	AMListAction3 *beamOnActionsList = new AMListAction3(new AMListActionInfo3("SXRMB Beam On", "SXRMB Beam On"), AMListAction3::Parallel);
-
-
+	// stage 1: open / wait the valves action list
 	AMListAction3 *valveOpenActionsList = new AMListAction3(new AMListActionInfo3("SXRMB Valve Open", "SXRMB Valve Open"), AMListAction3::Sequential);
+	AMListAction3 *valveWaitActionsList = new AMListAction3(new AMListActionInfo3("SXRMB Valve Wait", "SXRMB Valve Wait"), AMListAction3::Parallel);
 
-	AMControlInfo VVR16064B1003ValveSetpoint = VVR16064B1003Valve_->toInfo();
-	VVR16064B1003ValveSetpoint.setValue(1);
-	AMControlMoveAction3 *VVR16064B1003ValveSetpointOpenAction = new AMControlMoveAction3(new AMControlMoveActionInfo3(VVR16064B1003ValveSetpoint), VVR16064B1003Valve_);
-	valveOpenActionsList->addSubAction(VVR16064B1003ValveSetpointOpenAction);
+	if (VVR16064B1003Valve_->state() != 1) { // 1: OPEN 0: CLOSE 2: BETWEEN?
```

Comment:
Can yo use the CLSBiStateControl's getters of ```isOpen()``` ```isClosed()```?  It makes reading code like this much easier.